services/QuillLMS/app/services/classlink_integration/base_fetcher.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+module ClasslinkIntegration
+  class BaseFetcher < ApplicationService
+    DEFAULT_API_PAGE_SIZE = 10_000
+    attr_reader :district_oneroster_id, :district_bearer_token, :district_id
+
+    class NotImplementedError < StandardError; end
+
+    def run
+      transform_api_responses(fetch_all)
+    end
+
+    def initialize(district_oneroster_id:, district_bearer_token:, district_id:)
+      @district_oneroster_id = district_oneroster_id
+      @district_bearer_token = district_bearer_token
+      @district_id = district_id
+    end
+
+    # params:
+    #   resource_array: array of hashes, each hash representing a resource
+    # returns: array of hashes, sendable to AREL create methods
+    def transform_api_responses(resource_array)
+      resource_array.map do |api_response_fragment|
+        transform_api_response(api_response_fragment)
+      end
+    end
+
+    def transform_api_response(api_response_fragment)
+      raise NotImplementedError, 'You must implement this method in a subclass'
+    end
+
+    def resource_query_string = raise NotImplementedError
+
+    # sometimes, Classlink API responses are nested under a key that is not the resource name
+    # for example, the 'schools' resource is nested under the 'orgs' key
+    def api_response_selector_override = nil
+
+    # according to OneRoster spec, section 4.1, sourcedId is a GUID within a System
+    # in other words, it's the most globally unique identifier available
+    def fetch_all
+      resource = resource_query_string.split('?').first
+      api_response_selector = api_response_selector_override || resource
+
+      initial_page = fetch_page
+      total_item_count = initial_page.headers['x-total-count'].to_i
+      total_pages = (total_item_count / DEFAULT_API_PAGE_SIZE.to_f).ceil
```

Comment:
It adds a request, but you might consider doing a separate request / service that uses a limit of one in its request  to just calculate the number of pages (via the header info). That might clean up some of the code here so you could do something like:
```ruby
page_count = ClasslinkAPIPageCounter.run(district_oneroster_id:, resource:, query_string:)

results = (1..page_count).map {|page_number| fetch_page(page_number)[selector] }.flatten

Reply:
For now, I want to minimize requests, since Classlink mentions that it rate limits APIs, but does not advertise exact rate limits: https://help.classlink.com/s/article/rs-data-best-practices#complete